### ISSUE TITLE
Need help with debugging (IDFilter/MetaInfo)

### ISSUE THREAD

**Issue by hendrikweisser** [OpenMS/OpenMS#1760]
This is about an issue not with the current OpenMS code, but with a branch that I'm working on (to address #1706, i.e. refactor IDFilter):
https://github.com/hendrikweisser/OpenMS/tree/idfilter-extension

I'm getting a segmentation fault on one of the IDFilter TOPP tests, and I can't figure out what causes it. The affected test is "IDFilter_11", and the call that produces the problem is:
`IDFilter -in src/tests/topp/IDFilter_11_input.idXML -out IDFilter_11_output.tmp -remove_decoys`

The actual error depends on the architecture/compiler and the build type for OpenMS ("CMAKE_BUILD_TYPE"). I'm going to post some example stack traces below.
In a nutshell, some MetaInfo functions that look relatively innocuous to me produce segfaults in basic string operations.

If anyone is feeling adventurous/creative and would like to check out the branch to see if they can spot the error, I would really appreciate it.

**Comment by hendrikweisser** [comment]
Here's a GDB stack trace from a Debug build using "g++ (Ubuntu/Linaro 4.6.3-1ubuntu5) 4.6.3":

```
Starting program: /nfs/t17_project/OpenMS/build/bin/IDFilter -in /nfs/t17_project/OpenMS/src/tests/topp/IDFilter_11_input.idXML -out IDFilter_11_output.tmp -remove_decoys
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

---------------------------------------------------
FATAL: uncaught exception!
---------------------------------------------------
last entry in the exception handler:
exception of type OutOfMemory occured in line 72, function static void OpenMS::Exception::GlobalExceptionHandler::newHandler() of /nfs/t17_project/OpenMS/src/openms/source/CONCEPT/GlobalExceptionHandler.cpp
error message: unable to allocate enough memory (size = 0 bytes)
---------------------------------------------------

Program received signal SIGABRT, Aborted.
0x00007ffff50bc0d5 in __GI_raise (sig=<optimized out>) at ../nptl/sysdeps/unix/sysv/linux/raise.c:64
64      ../nptl/sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  0x00007ffff50bc0d5 in __GI_raise (sig=<optimized out>) at ../nptl/sysdeps/unix/sysv/linux/raise.c:64
#1  0x00007ffff50bf83b in __GI_abort () at abort.c:91
#2  0x00007ffff676cc44 in OpenMS::Exception::GlobalExceptionHandler::terminate () at /nfs/t17_project/OpenMS/src/openms/source/CONCEPT/GlobalExceptionHandler.cpp:110
#3  0x00007ffff592c846 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007ffff592b8d9 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007ffff592c16d in __gxx_personality_v0 () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00007ffff567071b in ?? () from /lib/x86_64-linux-gnu/libgcc_s.so.1
#7  0x00007ffff56709bd in _Unwind_RaiseException () from /lib/x86_64-linux-gnu/libgcc_s.so.1
#8  0x00007ffff592c961 in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#9  0x00007ffff676c9a3 in OpenMS::Exception::GlobalExceptionHandler::newHandler () at /nfs/t17_project/OpenMS/src/openms/source/CONCEPT/GlobalExceptionHandler.cpp:72
#10 0x00007ffff592ce0a in operator new(unsigned long) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#11 0x00007ffff5915a89 in std::string::_Rep::_S_create(unsigned long, unsigned long, std::allocator<char> const&) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#12 0x00007ffff59167fb in std::string::_Rep::_M_clone(std::allocator<char> const&, unsigned long) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#13 0x00007ffff5916f5c in std::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(std::string const&) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#14 0x00007ffff675d619 in OpenMS::String::String (this=0x7fffffff98a0) at /nfs/t17_project/OpenMS/src/openms/include/OpenMS/DATASTRUCTURES/String.h:57
#15 0x00007ffff676b04f in std::pair<OpenMS::String const, unsigned int>::pair (this=0x7fffffff98a0, __a=..., __b=@0x7fffffff98c8: 0) at /usr/include/c++/4.6/bits/stl_pair.h:104
#16 0x00007ffff676a818 in std::map<OpenMS::String, unsigned int, std::less<OpenMS::String>, std::allocator<std::pair<OpenMS::String const, unsigned int> > >::operator[] (this=0x7ffff7dd4768, __k=...)
    at /usr/include/c++/4.6/bits/stl_map.h:453
#17 0x00007ffff68f3918 in OpenMS::MetaInfoRegistry::registerName (this=0x7ffff7dd4760, name=..., description=..., unit=...) at /nfs/t17_project/OpenMS/src/openms/source/METADATA/MetaInfoRegistry.cpp:147
#18 0x00007ffff68f40d2 in OpenMS::MetaInfoRegistry::getIndex (this=0x7ffff7dd4760, name=...) at /nfs/t17_project/OpenMS/src/openms/source/METADATA/MetaInfoRegistry.cpp:246
#19 0x00007ffff68eeb7d in OpenMS::MetaInfo::exists (this=0x779da0, name=...) at /nfs/t17_project/OpenMS/src/openms/source/METADATA/MetaInfo.cpp:118
#20 0x00007ffff68f1cbb in OpenMS::MetaInfoInterface::metaValueExists (this=0xa11508, name=...) at /nfs/t17_project/OpenMS/src/openms/source/METADATA/MetaInfoInterface.cpp:143
#21 0x000000000042eec1 in OpenMS::IDFilter::HasMetaValue<OpenMS::ProteinHit>::operator() (this=0x7fffffff9af0, hit=...) at /nfs/t17_project/OpenMS/src/openms/include/OpenMS/FILTERING/ID/IDFilter.h:144
#22 0x000000000042d44d in OpenMS::IDFilter::HasDecoyAnnotation<OpenMS::ProteinHit>::operator() (this=0x7fffffff9af0, hit=...)
    at /nfs/t17_project/OpenMS/src/openms/include/OpenMS/FILTERING/ID/IDFilter.h:195
#23 0x000000000042ed1b in std::__find_if<__gnu_cxx::__normal_iterator<OpenMS::ProteinHit*, std::vector<OpenMS::ProteinHit, std::allocator<OpenMS::ProteinHit> > >, OpenMS::IDFilter::HasDecoyAnnotation<OpenMS::ProteinHit> > (__first=..., __last=..., __pred=...) at /usr/include/c++/4.6/bits/stl_algo.h:210
#24 0x000000000042d422 in std::find_if<__gnu_cxx::__normal_iterator<OpenMS::ProteinHit*, std::vector<OpenMS::ProteinHit, std::allocator<OpenMS::ProteinHit> > >, OpenMS::IDFilter::HasDecoyAnnotation<OpenMS::ProteinHit> > (__first=..., __last=..., __pred=...) at /usr/include/c++/4.6/bits/stl_algo.h:4427
#25 0x000000000042ab10 in std::remove_if<__gnu_cxx::__normal_iterator<OpenMS::ProteinHit*, std::vector<OpenMS::ProteinHit, std::allocator<OpenMS::ProteinHit> > >, OpenMS::IDFilter::HasDecoyAnnotation<OpenMS::ProteinHit> > (__first=..., __last=..., __pred=...) at /usr/include/c++/4.6/bits/stl_algo.h:1141
#26 0x00000000004285d3 in OpenMS::IDFilter::removeMatchingItems<std::vector<OpenMS::ProteinHit, std::allocator<OpenMS::ProteinHit> >, OpenMS::IDFilter::HasDecoyAnnotation<OpenMS::ProteinHit> > (items=...,
    pred=...) at /nfs/t17_project/OpenMS/src/openms/include/OpenMS/FILTERING/ID/IDFilter.h:239
#27 0x000000000042531c in OpenMS::IDFilter::filterIdentificationsByDecoy<OpenMS::ProteinIdentification> (identification=..., filtered_identification=...)
    at /nfs/t17_project/OpenMS/src/openms/include/OpenMS/FILTERING/ID/IDFilter.h:392
#28 0x00000000004225e0 in TOPPIDFilter::main_ (this=0x7fffffffda10) at /nfs/t17_project/OpenMS/src/topp/IDFilter.cpp:601
#29 0x00007ffff7112e63 in OpenMS::TOPPBase::main (this=0x7fffffffda10, argc=6, argv=0x7fffffffdf68) at /nfs/t17_project/OpenMS/src/openms/source/APPLICATIONS/TOPPBase.cpp:458
#30 0x000000000041c7b9 in main (argc=6, argv=0x7fffffffdf68) at /nfs/t17_project/OpenMS/src/topp/IDFilter.cpp:677
```

**Comment by hendrikweisser** [comment]
And here's one from a Debug build on Mac using "Apple LLVM version 5.0 (clang-500.2.79) (based on LLVM 3.3svn)":

```
Starting program: /Users/hw5/Software/OpenMS/build/bin/IDFilter -in /Users/hw5/Software/OpenMS/src/tests/topp/IDFilter_11_input.idXML -out IDFilter_11_output.tmp -remove_decoys
Reading symbols for shared libraries ++++++++.................................................................................................. done
Reading symbols for shared libraries . done

Program received signal EXC_BAD_ACCESS, Could not access memory.
Reason: KERN_INVALID_ADDRESS at address: 0x00007fff5fc84000
0x00007fff92073050 in memcmp ()
(gdb) bt
#0  0x00007fff92073050 in memcmp ()
#1  0x00007fff98199ff8 in std::string::compare ()
#2  0x0000000100014bad in std::operator< <char, std::char_traits<char>, std::allocator<char> > (__lhs=@0x1050f0fe0, __rhs=@0x7fff5fbfa550) at basic_string.h:2227
#3  0x0000000100014a84 in std::less<OpenMS::String>::operator() (this=0x1019fd118, __x=@0x1050f0fe0, __y=@0x7fff5fbfa550) at stl_function.h:227
#4  0x000000010029ee68 in std::_Rb_tree<OpenMS::String, std::pair<OpenMS::String const, unsigned int>, std::_Select1st<std::pair<OpenMS::String const, unsigned int> >, std::less<OpenMS::String>, std::allocator<std::pair<OpenMS::String const, unsigned int> > >::find (this=0x1019fd118, __k=@0x7fff5fbfa550) at stl_tree.h:1377
#5  0x000000010029e8bd in std::map<OpenMS::String, unsigned int, std::less<OpenMS::String>, std::allocator<std::pair<OpenMS::String const, unsigned int> > >::find (this=0x1019fd118, __x=@0x7fff5fbfa550) at stl_map.h:542
#6  0x000000010029d5d2 in OpenMS::MetaInfoRegistry::getIndex (this=0x1019fd110, name=@0x7fff5fbfa550) at /Users/hw5/Software/OpenMS/src/openms/source/METADATA/MetaInfoRegistry.cpp:237
#7  0x000000010029d667 in OpenMS::MetaInfoRegistry::getIndex (this=0x1019fd110, name=@0x7fff5fbfa550) at /Users/hw5/Software/OpenMS/src/openms/source/METADATA/MetaInfoRegistry.cpp:247
#8  0x0000000100296831 in OpenMS::MetaInfo::exists (this=0x10552ec80, name=@0x7fff5fbfa550) at /Users/hw5/Software/OpenMS/src/openms/source/METADATA/MetaInfo.cpp:118
#9  0x000000010029aafe in OpenMS::MetaInfoInterface::metaValueExists (this=0x1050323b8, name=@0x7fff5fbfa550) at /Users/hw5/Software/OpenMS/src/openms/source/METADATA/MetaInfoInterface.cpp:143
#10 0x000000010001239e in OpenMS::IDFilter::HasMetaValue<OpenMS::ProteinHit>::operator() (this=0x7fff5fbfa390, hit=@0x1050323b0) at IDFilter.h:144
#11 0x000000010001230f in OpenMS::IDFilter::HasDecoyAnnotation<OpenMS::ProteinHit>::operator() (this=0x7fff5fbfa390, hit=@0x1050323b0) at IDFilter.h:195
#12 0x0000000100012509 in std::__find_if<__gnu_cxx::__normal_iterator<OpenMS::ProteinHit*, std::vector<OpenMS::ProteinHit, std::allocator<OpenMS::ProteinHit> > >, OpenMS::IDFilter::HasDecoyAnnotation<OpenMS::ProteinHit> > (__first={_M_current = 0x1050323b0}, __last={_M_current = 0x105032560}, __pred={target_decoy = {key = @0x7fff5fbfa550, value = @0x7fff5fbfa530}, is_decoy = {key = @0x7fff5fbfa518, value = @0x7fff5fbfa4f8}}) at stl_algo.h:260
#13 0x0000000100012172 in std::find_if<__gnu_cxx::__normal_iterator<OpenMS::ProteinHit*, std::vector<OpenMS::ProteinHit, std::allocator<OpenMS::ProteinHit> > >, OpenMS::IDFilter::HasDecoyAnnotation<OpenMS::ProteinHit> > (__first={_M_current = 0x1050323b0}, __last={_M_current = 0x105032560}, __pred={target_decoy = {key = @0x7fff5fbfa550, value = @0x7fff5fbfa530}, is_decoy = {key = @0x7fff5fbfa518, value = @0x7fff5fbfa4f8}}) at stl_algo.h:349
#14 0x0000000100011f41 in std::remove_if<__gnu_cxx::__normal_iterator<OpenMS::ProteinHit*, std::vector<OpenMS::ProteinHit, std::allocator<OpenMS::ProteinHit> > >, OpenMS::IDFilter::HasDecoyAnnotation<OpenMS::ProteinHit> > (__first={_M_current = 0x1050323b0}, __last={_M_current = 0x105032560}, __pred={target_decoy = {key = @0x7fff5fbfa550, value = @0x7fff5fbfa530}, is_decoy = {key = @0x7fff5fbfa518, value = @0x7fff5fbfa4f8}}) at stl_algo.h:1298
#15 0x0000000100011dae in OpenMS::IDFilter::removeMatchingItems<std::vector<OpenMS::ProteinHit, std::allocator<OpenMS::ProteinHit> >, OpenMS::IDFilter::HasDecoyAnnotation<OpenMS::ProteinHit> > (items=@0xa5d000007fff5fbf, pred=@0xd99000007fff5fbf) at IDFilter.h:239
#16 0xa5b0000000010000 in ?? ()
```

**Comment by hendrikweisser** [comment]
Stack trace from an earlier RelWithDebInfo build on Linux:

```
Starting program: /nfs/t17_project/OpenMS/build/bin/IDFilter -in /nfs/t17_project/OpenMS/src/tests/topp/IDFilter_11_input.idXML -out IDFilter_11_output.tmp -remove_decoys
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
While loading '/nfs/t17_project/OpenMS/src/tests/topp/IDFilter_11_input.idXML': The XML file (1.3) is newer than the parser (1.2). This might lead to undefined program behavior.

Program received signal SIGSEGV, Segmentation fault.
__memcpy_ssse3_back () at ../sysdeps/x86_64/multiarch/memcpy-ssse3-back.S:1579
1579    ../sysdeps/x86_64/multiarch/memcpy-ssse3-back.S: No such file or directory.
(gdb) bt
#0  __memcpy_ssse3_back () at ../sysdeps/x86_64/multiarch/memcpy-ssse3-back.S:1579
#1  0x00007ffff6708842 in std::string::_Rep::_M_clone(std::allocator<char> const&, unsigned long) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#2  0x00007ffff6708f5c in std::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(std::string const&) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00007ffff6dce83c in String (this=0x7fffffff9790) at /nfs/t17_project/OpenMS/src/openms/include/OpenMS/DATASTRUCTURES/String.h:57
#4  OpenMS::MetaInfoRegistry::registerName (this=0x7ffff7dd3e60, name=..., description=..., unit=...) at /nfs/t17_project/OpenMS/src/openms/source/METADATA/MetaInfoRegistry.cpp:147
#5  0x00007ffff6dcee91 in OpenMS::MetaInfoRegistry::getIndex (this=0x7ffff7dd3e60, name=...) at /nfs/t17_project/OpenMS/src/openms/source/METADATA/MetaInfoRegistry.cpp:247
#6  0x00007ffff6dc9364 in OpenMS::MetaInfo::exists (this=0x6bf730, name=...) at /nfs/t17_project/OpenMS/src/openms/source/METADATA/MetaInfo.cpp:118
#7  0x000000000040e603 in operator() (hit=..., this=0x7fffffff9a50) at /nfs/t17_project/OpenMS/src/openms/include/OpenMS/FILTERING/ID/IDFilter.h:144
#8  operator() (hit=..., this=0x7fffffff9a50) at /nfs/t17_project/OpenMS/src/openms/include/OpenMS/FILTERING/ID/IDFilter.h:195
#9  std::__find_if<__gnu_cxx::__normal_iterator<OpenMS::ProteinHit*, std::vector<OpenMS::ProteinHit, std::allocator<OpenMS::ProteinHit> > >, OpenMS::IDFilter::HasDecoyAnnotation<OpenMS::ProteinHit> > (
    __first=..., __last=..., __pred=...) at /usr/include/c++/4.6/bits/stl_algo.h:210
#10 0x000000000040ed13 in find_if<__gnu_cxx::__normal_iterator<OpenMS::ProteinHit*, std::vector<OpenMS::ProteinHit> >, OpenMS::IDFilter::HasDecoyAnnotation<OpenMS::ProteinHit> > (__pred=..., __last=...,
    __first=...) at /usr/include/c++/4.6/bits/stl_algo.h:4427
#11 std::remove_if<__gnu_cxx::__normal_iterator<OpenMS::ProteinHit*, std::vector<OpenMS::ProteinHit, std::allocator<OpenMS::ProteinHit> > >, OpenMS::IDFilter::HasDecoyAnnotation<OpenMS::ProteinHit> > (
    __first=..., __last=..., __pred=...) at /usr/include/c++/4.6/bits/stl_algo.h:1141
#12 0x0000000000413232 in removeMatchingItems<std::vector<OpenMS::ProteinHit>, OpenMS::IDFilter::HasDecoyAnnotation<OpenMS::ProteinHit> > (pred=..., items=...)
    at /nfs/t17_project/OpenMS/src/openms/include/OpenMS/FILTERING/ID/IDFilter.h:239
#13 filterIdentificationsByDecoy<OpenMS::ProteinIdentification> (filtered_identification=..., identification=...) at /nfs/t17_project/OpenMS/src/openms/include/OpenMS/FILTERING/ID/IDFilter.h:392
#14 TOPPIDFilter::main_ (this=0x7fffffffda10) at /nfs/t17_project/OpenMS/src/topp/IDFilter.cpp:601
#15 0x00007ffff6cb97b5 in OpenMS::TOPPBase::main (this=0x7fffffffda10, argc=6, argv=0x7fffffffdf68) at /nfs/t17_project/OpenMS/src/openms/source/APPLICATIONS/TOPPBase.cpp:458
#16 0x0000000000406846 in main (argc=6, argv=0x7fffffffdf68) at /nfs/t17_project/OpenMS/src/topp/IDFilter.cpp:677
```

**Comment by timosachsenberg** [comment]
HI @hendrikweisser . const String& key; looks a bit dangerous. Are you sure that this never points to a e.g. temporary or string in a map that might be subject to reallocation before it is reused in operator()?

**Comment by timosachsenberg** [comment]
in:
```
    template <class HitType>
    struct HasMetaValue
    {
      typedef HitType argument_type; // for use as a predicate

      const String& key;
      const DataValue& value;

      HasMetaValue(const String& key, const DataValue& value):
        key(key), value(value)
      {}

      bool operator()(const HitType& hit) const
      {
        if (!hit.metaValueExists(key)) return false;
        if (value.isEmpty()) return true; // "key" is set, value doesn't matter
        DataValue found = hit.getMetaValue(key);
        // if (found.isEmpty()) return false; // meta value "key" not set
        // if (value.isEmpty()) return true; // "key" is set, value doesn't matter
        return found == value;
      }
    };
```

**Comment by hendrikweisser** [comment]
Thanks, Timo, that was quick and well spotted!
If I change the `const String&` to `String`, the crash goes away.

Now I've been thinking about why the compiler doesn't catch this - since usually it will complain when you try to initialize a const reference with a temporary value. I believe the underlying problem is in `MetaInfoRegistry::registerName` - which is declared const, but isn't actually const. The string reference is passed to that function from HasMetaValue, and I think that's where things actually go wrong, when the string variable that was referenced gets destroyed.

**Comment by hendrikweisser** [comment]
See also #1759 for problems with MetaInfoRegistry etc.

**Comment by hendrikweisser** [comment]
One thing that I just learned about C++:
In general, "whenever a reference is bound to a temporary, the lifetime of the temporary is extended to match the lifetime of the reference". That's as expected. Using the HasMetaValue class that Timo posted above as an example, I can write somewhere in my code:
`struct HasMetaValue<PeptideHit> is_decoy("target_decoy", "decoy");`, and it will be fine - although `"target_decoy"` and `"decoy"` are temporaries, their lifetime is extended and I can use them through references in `is_decoy`.

However, there are exceptions to the statement above, and the one that bit me is: "a temporary bound to a reference member in a constructor initializer list persists only until the constructor exits, not as long as the object exists"!
So if I make another class like this:
```
    template <class HitType>
    struct HasDecoyAnnotation
    {
      typedef HitType argument_type; // for use as a predicate

      struct HasMetaValue<HitType> is_decoy;

      HasDecoyAnnotation():
        is_decoy("target_decoy", "decoy")
      {}

      [...]
    }
```
Then I will run into big problems because the references in `is_decoy` will be invalid (because I initialized them with temporary variables in the initializer list of a constructor).

See here for more info: http://en.cppreference.com/w/cpp/language/reference_initialization

**Comment by timosachsenberg** [comment]
Nice to know... also note a potential source of ugly errors:
"In general, the lifetime of a temporary cannot be further extended by "passing it on": a second reference, initialized from the reference to which the temporary was bound, does not affect its lifetime."

So I guess best practice is to not reference temporaries at all (only use it for the usual cases of passing a temporary as const & parameter to a function).

**Comment by hendrikweisser** [comment]
> Nice to know... also note a potential source of ugly errors:
> "In general, the lifetime of a temporary cannot be further extended by "passing it on": a second reference, initialized from the reference to which the temporary was bound, does not affect its lifetime."

Yes, although in practice I don't see how that could happen unless you're dynamically creating objects (e.g. with "new"). If you want to initialize a "reference2" with a "reference1", "reference1" must already exist in the same or in a higher (previous) scope, so I don't see how it could go out of scope before "reference2".
